File: godtools-tract-renderer/src/main/java/org/cru/godtools/tract/model/Modal.java
package org.cru.godtools.tract.model;

import android.graphics.Color;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.View;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;

import org.ccci.gto.android.common.util.XmlPullParserUtils;
import org.cru.godtools.base.model.Event;
import org.cru.godtools.tract.R;
import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;

import java.io.IOException;
import java.util.List;
import java.util.Set;

import static org.cru.godtools.tract.Constants.XMLNS_TRACT;

public final class Modal extends Base implements Parent, StylesParent {
    static final String XML_MODAL = "modal";
    private static final String XML_TITLE = "title";
    private static final String XML_LISTENERS = "listeners";
    private static final String XML_DISMISS_LISTENERS = "dismiss-listeners";

    private final int mPosition;

    @NonNull
    private Set<Event.Id> mListeners = ImmutableSet.of();
    @NonNull
    private Set<Event.Id> mDismissListeners = ImmutableSet.of();

    @Nullable
    Text mTitle;

    @NonNull
    private List<Content> mContent = ImmutableList.of();

    private Modal(@NonNull final Base parent, final int position) {
        super(parent);
        mPosition = position;
    }

    @NonNull
    public String getId() {
        return getPage().getId() + "-" + mPosition;
    }

    @NonNull
    public Set<Event.Id> getListeners() {
        return mListeners;
    }

    @NonNull
    public Set<Event.Id> getDismissListeners() {
        return mDismissListeners;
    }

    @NonNull
    @Override
    public List<Content> getContent() {
        return mContent;
    }

    @Override
    public int getTextColor() {
        return Color.WHITE;
    }

    @Override
    public int getPrimaryColor() {
        return Color.TRANSPARENT;
    }

    @Override
    public int getPrimaryTextColor() {
        return Color.WHITE;
    }

    @Override
    public int getButtonColor() {
        return Color.WHITE;
    }

    @NonNull
    static Modal fromXml(@NonNull final Base parent, @NonNull final XmlPullParser parser, final int position)
            throws IOException, XmlPullParserException {
        return new Modal(parent, position).parse(parser);
    }

    @NonNull
    private Modal parse(@NonNull final XmlPullParser parser) throws IOException, XmlPullParserException {
        parser.require(XmlPullParser.START_TAG, XMLNS_TRACT, XML_MODAL);

        mListeners = parseEvents(parser, XML_LISTENERS);
        mDismissListeners = parseEvents(parser, XML_DISMISS_LISTENERS);

        // process any child elements
        final ImmutableList.Builder<Content> contentList = ImmutableList.builder();
        while (parser.next() != XmlPullParser.END_TAG) {
            if (parser.getEventType() != XmlPullParser.START_TAG) {
                continue;
            }

            // process recognized nodes
            switch (parser.getNamespace()) {
                case XMLNS_TRACT:
                    switch (parser.getName()) {
                        case XML_TITLE:
                            mTitle = Text.fromNestedXml(this, parser, XMLNS_TRACT, XML_TITLE);
                            continue;
                    }
                    break;
            }

            // try parsing this child element as a content node
            final Content content = Content.fromXml(this, parser);
            if (content != null) {
                contentList.add(content);
                continue;
            }

            // skip unrecognized nodes
            XmlPullParserUtils.skipTag(parser);
        }
        mContent = contentList.build();

        return this;
    }

    @NonNull
    public static ModalViewHolder getViewHolder(@NonNull final View root) {
        final Object holder = root.getTag(R.id.view_holder);
        if (holder instanceof ModalViewHolder) {
            return (ModalViewHolder) holder;
        } else {
            return new ModalViewHolder(root);
        }
    }

    public static class ModalViewHolder extends ParentViewHolder<Modal> {
        ModalViewHolder(@NonNull final View root) {
            super(Modal.class, root, null);
        }

        @Override
        void onBind() {
            super.onBind();
        }
    }
}

File: library/xml-model/src/main/java/org/cru/godtools/xml/model/Modal.kt
package org.cru.godtools.xml.model

import android.graphics.Color
import androidx.annotation.ColorInt
import androidx.annotation.DimenRes
import androidx.annotation.RestrictTo
import org.ccci.gto.android.common.util.XmlPullParserUtils
import org.cru.godtools.base.model.Event
import org.cru.godtools.xml.R
import org.cru.godtools.xml.XMLNS_TRACT
import org.cru.godtools.xml.model.Text.Companion.fromNestedXml
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserException
import java.io.IOException

private const val XML_TITLE = "title"

class Modal : Base, Parent, Styles {
    val id get() = "${page.id}-$position"
    private val position: Int

    val title: Text?
    override val content: List<Content>

    val listeners: Set<Event.Id>
    val dismissListeners: Set<Event.Id>

    @get:ColorInt
    override val primaryColor get() = Color.TRANSPARENT
    @get:ColorInt
    override val primaryTextColor get() = Color.WHITE
    @get:ColorInt
    override val textColor get() = Color.WHITE
    @get:ColorInt
    override val buttonColor get() = Color.WHITE

    @get:DimenRes
    override val textSize get() = R.dimen.text_size_modal
    override val textAlign get() = Text.Align.CENTER

    @RestrictTo(RestrictTo.Scope.TESTS)
    internal constructor(parent: Base, position: Int) : super(parent) {
        this.position = position
        title = null
        content = emptyList()
        listeners = emptySet()
        dismissListeners = emptySet()
    }

    @OptIn(ExperimentalStdlibApi::class)
    internal constructor(parent: Base, position: Int, parser: XmlPullParser) : super(parent) {
        this.position = position

        parser.require(XmlPullParser.START_TAG, XMLNS_TRACT, XML_MODAL)

        listeners = parseEvents(parser, XML_LISTENERS)
        dismissListeners = parseEvents(parser, XML_DISMISS_LISTENERS)

        // process any child elements
        var title: Text? = null
        content = buildList<Content> {
            parsingChildren@ while (parser.next() != XmlPullParser.END_TAG) {
                if (parser.eventType != XmlPullParser.START_TAG) continue

                when (parser.namespace) {
                    XMLNS_TRACT -> when (parser.name) {
                        XML_TITLE -> {
                            title = fromNestedXml(this@Modal, parser, XMLNS_TRACT, XML_TITLE)
                            continue@parsingChildren
                        }
                    }
                }

                // try parsing this child element as a content node
                val content = Content.fromXml(this@Modal, parser)
                if (content != null) {
                    if (!content.isIgnored) add(content)
                    continue
                }

                // skip unrecognized nodes
                XmlPullParserUtils.skipTag(parser)
            }
        }
        this.title = title
    }

    companion object {
        const val XML_MODAL = "modal"

        @JvmStatic
        @Deprecated(
            "Use constructor directly",
            ReplaceWith("Modal(parent, position, parser)", "org.cru.godtools.xml.model.Modal")
        )
        @Throws(IOException::class, XmlPullParserException::class)
        fun fromXml(parent: Base, parser: XmlPullParser, position: Int) = Modal(parent, position, parser)
    }
}
